File: BackEnd/src/agape/model/Evento.java
package agape.model;

import java.sql.Connection;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import agape.dao.EventoDAO;

public class Evento {
    private final EventoDAO dao = new EventoDAO();

    private int idEvento;
    private int idUsuarioResponsavel;
    private int idCatEvento;
    private String nome;
    private LocalDateTime dataInicio;
    private LocalDateTime dataFim;
    private int totVagas;
    private Integer vagasDisp;
    private Integer idEventoStatus;
    private LocalDateTime dataEvento;
    private LocalDateTime dataAberturaListaEspera;
    private Float valorInscricao;
    private String imagemEvento;

    // Campos extras (JOINs)
    private String nomeCatEvento;
    private String nomeStatus;

    public int getIdEvento() { return idEvento; }
    public void setIdEvento(int idEvento) { this.idEvento = idEvento; }

    public int getIdUsuarioResponsavel() { return idUsuarioResponsavel; }
    public void setIdUsuarioResponsavel(int idUsuarioResponsavel) { this.idUsuarioResponsavel = idUsuarioResponsavel; }

    public int getIdCatEvento() { return idCatEvento; }
    public void setIdCatEvento(int idCatEvento) { this.idCatEvento = idCatEvento; }

    public String getNome() { return nome; }
    public void setNome(String nome) { this.nome = nome; }

    public LocalDateTime getDataInicio() { return dataInicio; }
    public void setDataInicio(LocalDateTime dataInicio) { this.dataInicio = dataInicio; }

    public LocalDateTime getDataFim() { return dataFim; }
    public void setDataFim(LocalDateTime dataFim) { this.dataFim = dataFim; }

    public int getTotVagas() { return totVagas; }
    public void setTotVagas(int totVagas) { this.totVagas = totVagas; }

    public Integer getVagasDisp() { return vagasDisp; }
    public void setVagasDisp(Integer vagasDisp) { this.vagasDisp = vagasDisp; }

    public Integer getIdEventoStatus() { return idEventoStatus; }
    public void setIdEventoStatus(Integer idEventoStatus) { this.idEventoStatus = idEventoStatus; }

    public LocalDateTime getDataEvento() { return dataEvento; }
    public void setDataEvento(LocalDateTime dataEvento) { this.dataEvento = dataEvento; }

    public LocalDateTime getDataAberturaListaEspera() { return dataAberturaListaEspera; }
    public void setDataAberturaListaEspera(LocalDateTime dataAberturaListaEspera) { this.dataAberturaListaEspera = dataAberturaListaEspera; }

    public Float getValorInscricao() { return valorInscricao; }
    public void setValorInscricao(Float valorInscricao) { this.valorInscricao = valorInscricao; }

    public String getImagemEvento() { return imagemEvento; }
    public void setImagemEvento(String imagemEvento) { this.imagemEvento = imagemEvento; }

    public String getNomeCatEvento() { return nomeCatEvento; }
    public void setNomeCatEvento(String nomeCatEvento) { this.nomeCatEvento = nomeCatEvento; }

    public String getNomeStatus() { return nomeStatus; }
    public void setNomeStatus(String nomeStatus) { this.nomeStatus = nomeStatus; }

    // ── Persistência (DAO encapsulado) ─────────────────────────────────────

    public List<Evento> buscar(Connection conn, String nome, Integer idCatEvento, Integer idEventoStatus) throws Exception {
        return dao.buscar(conn, nome, idCatEvento, idEventoStatus);
    }

    public Evento buscarPorId(Connection conn, int id) throws Exception {
        return dao.buscarPorId(conn, id);
    }

    public void inserir(Connection conn) throws Exception {
        dao.inserir(conn, this);
    }

    public void atualizar(Connection conn) throws Exception {
        dao.atualizar(conn, this);
    }

    public void excluir(Connection conn) throws Exception {
        dao.excluir(conn, this.idEvento);
    }

    // ── Operações de estado (RF_F1) ────────────────────────────────────────

    /** novaData obrigatório ao reabrir cancelado; null para demais casos. */
    public void abrir(Connection conn, LocalDateTime novaData) throws Exception {
        dao.abrir(conn, this.idEvento, novaData);
        this.idEventoStatus = 1;
        if (novaData != null) this.dataEvento = novaData;
    }

    public void finalizar(Connection conn) throws Exception {
        dao.finalizar(conn, this.idEvento);
        this.idEventoStatus = 4;
    }

    public void cancelar(Connection conn) throws Exception {
        dao.cancelar(conn, this.idEvento);
        this.idEventoStatus = 2;
    }

    public void adiar(Connection conn, LocalDateTime novaData) throws Exception {
        dao.adiar(conn, this.idEvento, novaData);
        this.idEventoStatus = 3;
        this.dataEvento = novaData;
    }

    // ── Validações (SOLID SRP + GRASP Information Expert) ─────────────────
    // A entidade conhece suas próprias regras de negócio e as concentra aqui,
    // liberando o Controller de lógica de domínio.

    // ── Validação de disponibilidade para inscrição (RF_F3) ────────────────

    /**
     * GRASP Information Expert: a entidade Evento conhece seu próprio status
     * e determina se aceita novas inscrições.
     */
    public List<String> validarDisponibilidade() {
        List<String> erros = new ArrayList<>();
        if (idEventoStatus == null || idEventoStatus != 1)
            erros.add("Evento não está ativo para inscrições.");
        return erros;
    }

    /** Retorna true se ainda há vagas disponíveis. */
    public boolean temVagasDisponiveis() {
        return vagasDisp != null && vagasDisp > 0;
    }

    // ──────────────────────────────────────────────────────────────────────

    /** Valida campos obrigatórios para criar ou alterar um evento (Fluxo 7.1). */
    public List<String> validarParaCriar() {
        List<String> erros = new ArrayList<>();
        if (nome == null || nome.trim().length() < 2)
            erros.add("Nome deve ter pelo menos 2 caracteres.");
        if (idCatEvento <= 0)
            erros.add("Categoria do evento é obrigatória.");
        if (idUsuarioResponsavel <= 0)
            erros.add("Responsável pelo evento é obrigatório.");
        if (dataInicio == null)
            erros.add("Data de início das inscrições é obrigatória.");
        if (dataFim == null)
            erros.add("Data de término das inscrições é obrigatória.");
        if (dataInicio != null && dataFim != null && !dataFim.isAfter(dataInicio))
            erros.add("O término das inscrições deve ser posterior ao início.");
        if (dataFim != null && dataEvento != null && !dataEvento.isAfter(dataFim))
            erros.add("A data do evento deve ser posterior ao término das inscrições.");
        if (totVagas <= 0)
            erros.add("Total de vagas deve ser maior que zero.");
        return erros;
    }

    /** Valida abertura para eventos Adiados (sem nova data). */
    public List<String> validarParaAbrir() {
        List<String> erros = new ArrayList<>();
        if (idEventoStatus != null && idEventoStatus == 1)
            erros.add("Evento já está ativo.");
        if (idEventoStatus != null && idEventoStatus == 4)
            erros.add("Evento finalizado não pode ser reaberto.");
        return erros;
    }

    /** Valida reabertura de evento Cancelado: exige nova data futura. */
    public List<String> validarParaReabrirCancelado(LocalDateTime novaData) {
        List<String> erros = new ArrayList<>();
        if (novaData == null)
            erros.add("É necessário informar uma nova data para reabrir um evento cancelado.");
        if (novaData != null && novaData.isBefore(LocalDateTime.now()))
            erros.add("A nova data do evento deve ser uma data futura.");
        return erros;
    }

    /** Valida a transição de status ao finalizar um evento. */
    public List<String> validarParaFinalizar() {
        List<String> erros = new ArrayList<>();
        if (idEventoStatus != null && idEventoStatus == 4)
            erros.add("Evento já está finalizado.");
        if (idEventoStatus != null && idEventoStatus == 2)
            erros.add("Evento cancelado não pode ser finalizado diretamente. Reabra-o primeiro.");
        return erros;
    }

    /** Valida a transição de status ao cancelar um evento. */
    public List<String> validarParaCancelar() {
        List<String> erros = new ArrayList<>();
        if (idEventoStatus != null && idEventoStatus == 2)
            erros.add("Evento já está cancelado.");
        if (idEventoStatus != null && idEventoStatus == 4)
            erros.add("Evento finalizado não pode ser cancelado.");
        return erros;
    }

    /** Valida o adiamento: nova data deve ser futura e após o início. */
    public List<String> validarParaAdiar(LocalDateTime novaData) {
        List<String> erros = new ArrayList<>();
        if (novaData == null)
            erros.add("Nova data do evento é obrigatória para adiar.");
        if (novaData != null && novaData.isBefore(LocalDateTime.now()))
            erros.add("A nova data do evento deve ser uma data futura.");
        return erros;
    }

    public String toJson() {
        return "{" +
            "\"idEvento\":"                   + idEvento                      + "," +
            "\"idUsuarioResponsavel\":"        + idUsuarioResponsavel          + "," +
            "\"idCatEvento\":"                 + idCatEvento                   + "," +
            "\"nomeCatEvento\":"               + str(nomeCatEvento)            + "," +
            "\"idEventoStatus\":"              + (idEventoStatus != null ? idEventoStatus : "null") + "," +
            "\"nomeStatus\":"                  + str(nomeStatus)               + "," +
            "\"nome\":"                        + str(nome)                     + "," +
            "\"dataInicio\":"                  + dt(dataInicio)                + "," +
            "\"dataFim\":"                     + dt(dataFim)                   + "," +
            "\"totVagas\":"                    + totVagas                      + "," +
            "\"vagasDisp\":"                   + (vagasDisp != null ? vagasDisp : "null") + "," +
            "\"dataEvento\":"                  + dt(dataEvento)                + "," +
            "\"dataAberturaListaEspera\":"     + dt(dataAberturaListaEspera)   + "," +
            "\"valorInscricao\":"              + (valorInscricao != null ? valorInscricao : "null") + "," +
            "\"imagemEvento\":"                + str(imagemEvento)             +
        "}";
    }

    private String str(String s) {
        if (s == null) return "null";
        return "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
    }

    private String dt(LocalDateTime d) {
        return d != null ? "\"" + d.toString() + "\"" : "null";
    }
}

File: FrontEnd/models/Evento.js
/* Evento — entidade com validações e serialização para o backend */

window.AGAPE = window.AGAPE || {};
window.AGAPE.Models = window.AGAPE.Models || {};

window.AGAPE.Models.Evento = (function () {

    function Evento(dados) {
        dados = dados || {};
        this._idEvento                 = dados.idEvento                 || null;
        this._idUsuarioResponsavel     = dados.idUsuarioResponsavel     || null;
        this._idCatEvento              = dados.idCatEvento              || null;
        this._nome                     = dados.nome                     || '';
        this._dataInicio               = dados.dataInicio               || '';
        this._dataFim                  = dados.dataFim                  || '';
        this._totVagas                 = dados.totVagas != null ? Number(dados.totVagas) : null;
        this._vagasDisp                = dados.vagasDisp != null ? Number(dados.vagasDisp) : null;
        this._idEventoStatus           = dados.idEventoStatus           || 1;
        this._dataEvento               = dados.dataEvento               || '';
        this._dataAberturaListaEspera  = dados.dataAberturaListaEspera  || null;
        this._valorInscricao           = dados.valorInscricao != null ? dados.valorInscricao : null;
        this._imagemEvento             = dados.imagemEvento             || null;
        this._imagemBase64             = dados.imagemBase64             || null;
        this._nomeImagemEvento         = dados.nomeImagemEvento         || null;
    }

    Evento.prototype.getIdEvento                = function () { return this._idEvento;                };
    Evento.prototype.getIdUsuarioResponsavel    = function () { return this._idUsuarioResponsavel;    };
    Evento.prototype.getIdCatEvento             = function () { return this._idCatEvento;             };
    Evento.prototype.getNome                    = function () { return this._nome;                    };
    Evento.prototype.getDataInicio              = function () { return this._dataInicio;              };
    Evento.prototype.getDataFim                 = function () { return this._dataFim;                 };
    Evento.prototype.getTotVagas                = function () { return this._totVagas;                };
    Evento.prototype.getVagasDisp               = function () { return this._vagasDisp;               };
    Evento.prototype.getIdEventoStatus          = function () { return this._idEventoStatus;          };
    Evento.prototype.getDataEvento              = function () { return this._dataEvento;              };
    Evento.prototype.getDataAberturaListaEspera = function () { return this._dataAberturaListaEspera; };
    Evento.prototype.getValorInscricao          = function () { return this._valorInscricao;          };
    Evento.prototype.getImagemEvento            = function () { return this._imagemEvento;            };

    Evento.prototype.setIdEvento                = function (v) { this._idEvento                = v; };
    Evento.prototype.setIdUsuarioResponsavel    = function (v) { this._idUsuarioResponsavel    = v; };
    Evento.prototype.setIdCatEvento             = function (v) { this._idCatEvento             = v; };
    Evento.prototype.setNome                    = function (v) { this._nome                    = v; };
    Evento.prototype.setDataInicio              = function (v) { this._dataInicio              = v; };
    Evento.prototype.setDataFim                 = function (v) { this._dataFim                 = v; };
    Evento.prototype.setTotVagas                = function (v) { this._totVagas                = v != null ? Number(v) : null; };
    Evento.prototype.setVagasDisp               = function (v) { this._vagasDisp               = v != null ? Number(v) : null; };
    Evento.prototype.setIdEventoStatus          = function (v) { this._idEventoStatus          = v; };
    Evento.prototype.setDataEvento              = function (v) { this._dataEvento              = v; };
    Evento.prototype.setDataAberturaListaEspera = function (v) { this._dataAberturaListaEspera = v || null; };
    Evento.prototype.setValorInscricao          = function (v) { this._valorInscricao          = v || null; };
    Evento.prototype.setImagemEvento            = function (v) { this._imagemEvento            = v || null; };
    Evento.prototype.setImagemBase64            = function (v) { this._imagemBase64            = v || null; };
    Evento.prototype.setNomeImagemEvento        = function (v) { this._nomeImagemEvento        = v || null; };

    Evento.prototype.validar = function () {
        var erros = [];
        if (!this._nome || this._nome.trim().length < 2)
            erros.push('Nome deve ter pelo menos 2 caracteres.');
        if (!this._idCatEvento)
            erros.push('Selecione uma categoria.');
        if (!this._idUsuarioResponsavel)
            erros.push('Selecione o responsável.');
        if (!this._dataInicio)
            erros.push('Informe a data de início.');
        if (!this._dataFim)
            erros.push('Informe a data de término.');
        if (!this._dataEvento)
            erros.push('Informe a data do evento.');
        if (this._totVagas === null || this._totVagas <= 0)
            erros.push('Total de vagas deve ser maior que zero.');
        return erros;
    };

    Evento.prototype.paraJson = function () {
        var dados = {
            nome:                    this._nome,
            idCatEvento:             this._idCatEvento,
            idUsuarioResponsavel:    this._idUsuarioResponsavel,
            dataInicio:              this._dataInicio,
            dataFim:                 this._dataFim,
            dataEvento:              this._dataEvento,
            totVagas:                this._totVagas,
            vagasDisp:               this._vagasDisp,
            idEventoStatus:          this._idEventoStatus
        };
        if (this._idEvento !== null && this._idEvento !== undefined) dados.idEvento = this._idEvento;
        if (this._dataAberturaListaEspera) dados.dataAberturaListaEspera = this._dataAberturaListaEspera;
        if (this._valorInscricao != null)  dados.valorInscricao          = this._valorInscricao;
        if (this._imagemBase64)            dados.imagemBase64            = this._imagemBase64;
        if (this._nomeImagemEvento)        dados.nomeImagemEvento        = this._nomeImagemEvento;
        return dados;
    };

    // Mantido por compatibilidade
    Evento.prototype.paraFormData = Evento.prototype.paraJson;

    return Evento;

})();
